source/extensions/load_balancing_policies/client_side_weighted_round_robin/client_side_weighted_round_robin_lb.cc

```diff
@@ -0,0 +1,305 @@
+    updateWeightsOnHosts(host_set->hosts());
+  }
+}
+
+void ClientSideWeightedRoundRobinLoadBalancer::updateWeightsOnHosts(const HostVector& hosts) {
+  std::vector<uint32_t> weights;
+  HostVector hosts_with_default_weight;
+  const MonotonicTime now = time_source_.monotonicTime();
+  // Weight is considered invalid (too recent) if it was first updated within `blackout_period_`.
+  const MonotonicTime min_non_empty_since = now - blackout_period_;
+  // Weight is considered invalid (too old) if it was last updated before
+  // `weight_expiration_period_`.
+  const MonotonicTime max_last_update_time = now - weight_expiration_period_;
+  weights.reserve(hosts.size());
+  hosts_with_default_weight.reserve(hosts.size());
+  ENVOY_LOG(trace, "updateWeights hosts.size() = {}, time since epoch = {}", hosts.size(),
+            now.time_since_epoch().count());
+  // Scan through all hosts and update their weights if they are valid.
+  for (const auto& host_ptr : hosts) {
+    // Get client side weight or `nullopt` if it is invalid (see above).
+    absl::optional<uint32_t> client_side_weight =
+        getClientSideWeightIfValidFromHost(*host_ptr, min_non_empty_since, max_last_update_time);
+    // If `client_side_weight` is valid, then set it as the host weight and store it in
+    // `weights` to calculate median valid weight across all hosts.
+    if (client_side_weight.has_value()) {
+      weights.push_back(*client_side_weight);
+      host_ptr->weight(*client_side_weight);
+      ENVOY_LOG(trace, "updateWeights hostWeight {} = {}", getHostAddress(host_ptr.get()),
+                host_ptr->weight());
+    } else {
+      // If `client_side_weight` is invalid, then set host to default (median) weight.
+      hosts_with_default_weight.push_back(host_ptr);
+    }
+  }
+  // Calculate the default weight as median of all valid weights.
+  uint32_t default_weight = 1;
+  if (!weights.empty()) {
+    auto median_it = weights.begin() + weights.size() / 2;
+    std::nth_element(weights.begin(), median_it, weights.end());
+    default_weight = *median_it;
+  }
+  // Update the hosts with default weight.
+  for (const auto& host_ptr : hosts_with_default_weight) {
+    host_ptr->weight(default_weight);
+    ENVOY_LOG(trace, "updateWeights default hostWeight {} = {}", getHostAddress(host_ptr.get()),
+              host_ptr->weight());
+  }
+}
+
+void ClientSideWeightedRoundRobinLoadBalancer::addClientSideLbPolicyDataToHosts(
+    const HostVector& hosts) {
+  for (const auto& host_ptr : hosts) {
+    if (host_ptr->lbPolicyData() == nullptr) {
+      ENVOY_LOG(trace, "Adding LB policy data to Host {}", getHostAddress(host_ptr.get()));
+      host_ptr->setLbPolicyData(std::make_shared<ClientSideHostLbPolicyData>());
+    }
+  }
+}
+
+absl::optional<uint32_t>
+ClientSideWeightedRoundRobinLoadBalancer::getClientSideWeightIfValidFromHost(
+    const Host& host, const MonotonicTime& min_non_empty_since,
+    const MonotonicTime& max_last_update_time) {
```

Comment:
The naming of the parameter is confusing. See, the entries < max_last_update_time will be treat as invalid. So, the name should be `min_last_update_time`. Older then the min_last_update_time will be ignore. The `min_non_empty_since` is similar case.